tools/palette.py

```diff
@@ -0,0 +1,645 @@
+        raise Exception(f"ERROR: Parsed color count does not align with supported NAOMI palette sizes! {output.color_count} != 16 or 256")
+
+    output.raw_source = "".join(lines)
+
+    return output
+
+# Parse a .TXT palette file and return colors in the form of r,g,b,a values.
+# Format spec seems to be: https://www.getpaint.net/doc/latest/WorkingWithPalettes.html
+def parse_txt(input: UnparsedPaletteData) -> ParsedPaletteData:
+    output = ParsedPaletteData()
+    output.palette_type = input.palette_type
+
+    with open(input.args.palette, "rt") as f:
+        lines = f.readlines()
+
+        for index, line in enumerate(lines, start=0):
+            line = line.strip()
+            
+            # Make sure this is a valid Paint.net TXT format file
+            if index == 0:
+                if line.lower().find("paint.net palette file") != -1:
+                    continue
+                elif input.args.force == False:
+                    raise Exception(f"ERROR: Invalid or malformed TXT palette file passed to palette.py! {input.args.palette}")
+
+            # Parse TXT format
+            # Name is not part of spec, but may be included
+            if line.lower().find("name:") != -1:
+                nameSplit = line.split(":")
+                output.palette_name = nameSplit[-1]
+                continue
+
+            # Color count is not part of spec, but may be included
+            if line.lower().find("colors:") != -1:
+                color_countSplit = line.split(":")
+                output.color_count = int(color_countSplit[-1])
+                continue
+
+            # Ignore other commented lines
+            if line.find(";") != -1:
+                continue
+            
+            # Colors are listed in hex format aarrggbb
+            if len(line) >= 8:
+                output.palette_colors.append([int(line[2:4],16), int(line[4:6],16), int(line[6:8],16), int(line[0:2],16)])
+    
+    if output.color_count == -1:
+        output.color_count = len(output.palette_colors)
+
+    output.raw_source = "".join(lines)
+
+    return output
+
+# Parse a .PNG file and return return colors in the form of r,g,b,a values.
+# Processes the first 16 or 256 pixels of the PNG.
+def parse_png(input: UnparsedPaletteData) -> ParsedPaletteData:
+    output = ParsedPaletteData()
+    output.palette_type = input.palette_type
+
+    # Open image, get pixels array, convert to RGBA
+    with open(input.args.palette, "rb") as bfp:
+        data = bfp.read()
+    texture = Image.open(io.BytesIO(data))
+    width, height = texture.size
+
+    # Determine the number of pixels to encode as a palette
+    # Automatically take the larger possible size if possible
+    pix_count = width * height
+    if pix_count < 16:
+        raise Exception(f"ERROR: PNG pixel count is smaller than minimum NAOMI palette size! {width * height} < 16")
+    elif pix_count >= 16 and pix_count < 256:
+        pix_count = 16
+    else:
+        pix_count = 256
```

Comment:
Same comment here, it seems we should probably just size the palette up to the next increment (16 or 256) and only throw if its larger than 256.

Reply:
Sure thing, can do. I was actually waffling back and forth between rejecting or resizing off-size palettes when I was implementing this; both solutions have some downsides, but I think you're right that upsizing (and maybe throwing a warning if it upsizes from 16 -> 256, since that's a pretty significant difference in terms of actual use in a project) is the friendlier option.